source/common/stats/stats_impl.h

```diff
@@ -373,30 +377,58 @@ class SourceImpl : public Source {
 };
 
 /**
- * Implementation of RawStatDataAllocator that uses an unordered set to store
- * RawStatData pointers.
+ * This structure is an alternate backing store for both CounterImpl and GaugeImpl. It is designed
+ * so that it can be allocated efficiently from the heap on demand.
  */
-class HeapRawStatDataAllocator : public RawStatDataAllocator {
+struct HeapStatData {
+  explicit HeapStatData(absl::string_view key);
+
+  /**
+   * @returns absl::string_view the name as a string_view.
+   */
+  absl::string_view key() const { return name_; }
+
+  std::atomic<uint64_t> value_{0};
+  std::atomic<uint64_t> pending_increment_{0};
+  std::atomic<uint16_t> flags_{0};
+  std::atomic<uint16_t> ref_count_{1};
+  std::string name_;
+};
+
+/**
+ * Implementation of StatDataAllocator using a pure heap-based strategy, so that
+ * Envoy implementations that do not require hot-restart can use less memory.
+ */
+class HeapStatDataAllocator : public StatDataAllocatorImpl<HeapStatData> {
 public:
-  // RawStatDataAllocator
-  ~HeapRawStatDataAllocator() { ASSERT(stats_.empty()); }
-  RawStatData* alloc(const std::string& name) override;
-  void free(RawStatData& data) override;
+  HeapStatDataAllocator() {}
+  ~HeapStatDataAllocator() { ASSERT(stats_.empty()); }
+
+  // StatDataAllocatorImpl
+  HeapStatData* alloc(absl::string_view name) override;
+  void free(HeapStatData& data) override;
+
+  // StatsDataAllocator
+  bool requiresBoundedStatNameSize() const override { return false; }
 
 private:
-  struct RawStatDataHash_ {
-    size_t operator()(const RawStatData* a) const { return HashUtil::xxHash64(a->key()); }
+  struct HeapStatHash_ {
+    size_t operator()(const HeapStatData* a) const { return HashUtil::xxHash64(a->key()); }
   };
-  struct RawStatDataCompare_ {
-    bool operator()(const RawStatData* a, const RawStatData* b) const {
+  struct HeapStatCompare_ {
+    bool operator()(const HeapStatData* a, const HeapStatData* b) const {
       return (a->key() == b->key());
     }
   };
-  typedef std::unordered_set<RawStatData*, RawStatDataHash_, RawStatDataCompare_> StringRawDataSet;
 
-  // An unordered set of RawStatData pointers which keys off the key()
+  // TODO(jmarantz): See https://github.com/envoyproxy/envoy/pull/3927 which can
```

Comment:
Sounds like more of a TODO for @ambuc :)

Reply:
In my experience, TODO's are generally signed by the author of the TODO, not necessarily by the person who is expected to do it.

Reply:
Let's leave a comment linking to issue https://github.com/envoyproxy/envoy/issues/3585 which will address this in #3927, hopefully.